src/primitives/gf32.rs

```diff
@@ -0,0 +1,470 @@
+    /// Numeric value maps to bech32 character: 12 == "v".
+    pub const V: Fe32 = Fe32(12);
+    /// Numeric value maps to bech32 character: 13 == "d".
+    pub const D: Fe32 = Fe32(13);
+    /// Numeric value maps to bech32 character: 14 == "w".
+    pub const W: Fe32 = Fe32(14);
+    /// Numeric value maps to bech32 character: 15 == "0".
+    pub const _0: Fe32 = Fe32(15);
+    /// Numeric value maps to bech32 character: 16 == "s".
+    pub const S: Fe32 = Fe32(16);
+    /// Numeric value maps to bech32 character: 17 == "3".
+    pub const _3: Fe32 = Fe32(17);
+    /// Numeric value maps to bech32 character: 18 == "j".
+    pub const J: Fe32 = Fe32(18);
+    /// Numeric value maps to bech32 character: 19 == "n".
+    pub const N: Fe32 = Fe32(19);
+    /// Numeric value maps to bech32 character: 20 == "5".
+    pub const _5: Fe32 = Fe32(20);
+    /// Numeric value maps to bech32 character: 21 == "4".
+    pub const _4: Fe32 = Fe32(21);
+    /// Numeric value maps to bech32 character: 22 == "k".
+    pub const K: Fe32 = Fe32(22);
+    /// Numeric value maps to bech32 character: 23 == "h".
+    pub const H: Fe32 = Fe32(23);
+    /// Numeric value maps to bech32 character: 24 == "c".
+    pub const C: Fe32 = Fe32(24);
+    /// Numeric value maps to bech32 character: 25 == "e".
+    pub const E: Fe32 = Fe32(25);
+    /// Numeric value maps to bech32 character: 26 == "6".
+    pub const _6: Fe32 = Fe32(26);
+    /// Numeric value maps to bech32 character: 27 == "m".
+    pub const M: Fe32 = Fe32(27);
+    /// Numeric value maps to bech32 character: 28 == "u".
+    pub const U: Fe32 = Fe32(28);
+    /// Numeric value maps to bech32 character: 29 == "a".
+    pub const A: Fe32 = Fe32(29);
+    /// Numeric value maps to bech32 character: 30 == "7".
+    pub const _7: Fe32 = Fe32(30);
+    /// Numeric value maps to bech32 character: 31 == "l".
+    pub const L: Fe32 = Fe32(31);
+
+    /// Iterator over all field elements, in alphabetical order.
+    pub fn iter_alpha() -> impl Iterator<Item = Fe32> {
+        [
+            Fe32::A,
+            Fe32::C,
+            Fe32::D,
+            Fe32::E,
+            Fe32::F,
+            Fe32::G,
+            Fe32::H,
+            Fe32::J,
+            Fe32::K,
+            Fe32::L,
+            Fe32::M,
+            Fe32::N,
+            Fe32::P,
+            Fe32::Q,
+            Fe32::R,
+            Fe32::S,
+            Fe32::T,
+            Fe32::U,
+            Fe32::V,
+            Fe32::W,
+            Fe32::X,
+            Fe32::Y,
+            Fe32::Z,
+            Fe32::_0,
+            Fe32::_2,
+            Fe32::_3,
+            Fe32::_4,
+            Fe32::_5,
+            Fe32::_6,
+            Fe32::_7,
+            Fe32::_8,
+            Fe32::_9,
+        ]
+        .iter()
+        .copied()
+    }
+
+    /// Creates a field element from a single bech32 character.
+    pub fn from_char(c: char) -> Result<Fe32, Error> {
+        let ascii = u32::from(c) as usize;
```

Comment:
In 45b7cd8:

This cast might truncate. Using `usize::try_from` would be better. Or you could use `i8::try_from`, like the original code did, followed by `usize::try_from`, if you want to be fancy.

Reply:
Can we trick the test suite into giving us 16-bit `usize`?

Reply:
That would be cool, but no, not that I'm aware of.

Reply:
I thought we didn't support 16 bit machines? Want me to add
```rust
// Disable 16-bit support at least for now as we can't guarantee it yet.
#[cfg(target_pointer_width = "16")]
compile_error!(
    "rust-bitcoin currently only supports architectures with pointers wider than 16 bits, let us
    know if you want 16-bit support. Note that we do NOT guarantee that we will implement it!"
);
```

??

Reply:
With `s/rust-bitcoin/rust-bech32/`

Reply:
No, I think we should just support 16-bit machines, since it has zero cost for this crate.

Reply:
I'm not sure this is much better than the original but I've left it as:
```rust
    /// Creates a field element from a single bech32 character.
    pub fn from_char(c: char) -> Result<Fe32, Error> {
        // i8::try_from gets a value in the range 0..=127 since char is unsigned.
        let byte = i8::try_from(u32::from(c)).map_err(|_| Error::InvalidChar(c))?;
        let ascii = usize::try_from(byte).expect("byte value within range 0..=127");
        // We use -1 for any array element that is an invalid char to trigger error from u8::try_from
        let u5 = u8::try_from(CHARS_INV[ascii]).map_err(|_| Error::InvalidChar(c))?;
        Ok(Fe32(u5))
    }
```